src/core/thread/link_metrics_tlvs.hpp

```diff
@@ -0,0 +1,332 @@
+    bool IsExtendedFlagSet(void) const { return (mTypeIdFlags & kExtendedFlag) != 0; }
+
+    /**
+     * This method clears value length flag.
+     *
+     */
+    void ClearLengthFlag(void) { mTypeIdFlags &= ~kLengthFlag; }
+
+    /**
+     * This method sets the value length flag.
+     *
+     */
+    void SetLengthFlag(void) { mTypeIdFlags |= kLengthFlag; }
+
+    /**
+     * This method indicates whether or not the value length flag is set.
+     *
+     * @retval true   The value length flag is set, extended value length (4 bytes)
+     * @retval false  The value length flag is not set, short value length (1 byte)
+     */
+    bool IsLengthFlagSet(void) const { return (mTypeIdFlags & kLengthFlag) != 0; }
+
+    /**
+     * This method sets the Type/Average Enum.
+     *
+     * @param[in]  aTypeEnum  Type/Average Enum.
+     *
+     */
+    void SetTypeEnum(uint8_t aTypeEnum)
+    {
+        mTypeIdFlags = (mTypeIdFlags & ~kTypeEnumMask) | ((aTypeEnum << kTypeEnumOffset) & kTypeEnumMask);
+    }
+
+    /**
+     * This method returns the Type/Average Enum.
+     *
+     * @returns The Type/Average Enum.
+     *
+     */
+    uint8_t GetTypeEnum(void) const { return (mTypeIdFlags & kTypeEnumMask) >> kTypeEnumOffset; }
+
+    /**
+     * This method sets the Metric Enum.
+     *
+     * @param[in]  aMetricEnum  Metric Enum.
+     *
+     */
+    void SetMetricEnum(uint8_t aMetricEnum)
+    {
+        mTypeIdFlags = (mTypeIdFlags & ~kMetricEnumMask) | ((aMetricEnum << kMetricEnumOffset) & kMetricEnumMask);
+    }
+
+    /**
+     * This method returns the Metric Enum.
+     *
+     * @returns The Metric Enum.
+     *
+     */
+    uint8_t GetMetricEnum(void) const { return (mTypeIdFlags & kMetricEnumMask) >> kMetricEnumOffset; }
+
+    /**
+     * This method returns the raw value of the entire TypeIdFlags.
+     *
+     * @returns The raw value of TypeIdFlags.
+     *
+     */
+    uint8_t GetRawValue(void) const { return mTypeIdFlags; }
+
+private:
+    enum
+    {
+        kLengthFlag       = 1 << 6,
+        kExtendedFlag     = 1 << 7,
+        kTypeEnumOffset   = 3,
+        kTypeEnumMask     = 7 << kTypeEnumOffset,
+        kMetricEnumOffset = 0,
+        kMetricEnumMask   = 7 << kMetricEnumOffset,
+    };
+
+    uint8_t mTypeIdFlags;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Link Metrics Report Sub-TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class LinkMetricsReportSubTlv : public Tlv
+{
+public:
+    /**
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void)
+    {
+        SetType(kLinkMetricsReportSub);
+        SetLength(sizeof(*this) - sizeof(Tlv));
+    }
+
+    /**
+     * This method indicates whether or not the TLV appears to be well-formed.
+     *
+     * @retval true   The TLV appears to be well-formed.
+     * @retval false  The TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const { return GetLength() == sizeof(*this) - sizeof(Tlv); }
```

Comment:
I think I mentioned this before.

We should always accept TLVs (consider it valid) if theie length is larger than some min length we know (based on what current spec defines).We should not use exact length check (or check length is smaller than somthing).

BTW, this TLV can have different lengths based on `L` flag in the Type Id (whether the metric value is  `m8` or `m32`).
The check above inherently assumes `m32` due to `union` definition below.